python/pyspark/ml/functions.py

```diff
@@ -106,6 +111,167 @@ def array_to_vector(col: Column) -> Column:
     return Column(sc._jvm.org.apache.spark.ml.functions.array_to_vector(_to_java_column(col)))
 
 
+def batched(df: pd.DataFrame, batch_size: int = -1) -> Iterator[pd.DataFrame]:
+    """Generator that splits a pandas dataframe/series into batches."""
+    if batch_size <= 0 or batch_size >= len(df):
+        yield df
+    else:
+        # for batch in np.array_split(df, (len(df.index) + batch_size - 1) // batch_size):
+        for _, batch in df.groupby(np.arange(len(df)) // batch_size):
+            yield batch
+
+
+def has_tensor_cols(df: pd.DataFrame) -> bool:
+    """Check if input DataFrame contains any tensor-valued columns"""
+    if any(df.dtypes == np.object_):
+        # pd.DataFrame object types can contain different types, e.g. string, dates, etc.
+        # so inspect a row and check for array/list type
+        sample = df.iloc[0]
+        return any([isinstance(x, np.ndarray) or isinstance(x, list) for x in sample])
+    else:
+        return False
+
+
+def batch_infer_udf(
+    predict_batch_fn: Callable,
+    return_type: DataType = ArrayType(FloatType()),
+    batch_size: int = -1,
+    input_names: list[str] = [],
```

Comment:
* Shouldn't use mutable `[]` as default value. Use `None` instead.
* Not sure if we need this param in the first version. We can restrict the input to either:
** One value column (no name), `model_udf(c1)`
** A struct column (caller can rename the sub-columns if needed). `model_udf(struct(c1, c2))`. Or another option is to accept a list of columns `(model_udf(c1, c2))`.  `predict_batch_fn` and `input_tensor_shapes` signature should match accordingly. I slightly prefer the latter which doesn't need to handle name conflicts and we can use list for tensor shapes.

Reply:
- mutable `[]` is fixed.
- for now, the `input_names` is an explicit signal to the code that the user is expecting multiple named inputs, and for most simple (single-input) cases, this wouldn't be used/required.

Reply:
Per offline discussion with @mengxr, removed `input_names` and dictionary-style inputs in favor of list-style inputs.